File: packages/alpaca-option/src/math/geometricAsian.ts
import { fail } from '../error';
import { normalCdf } from '../numeric';
import type { OptionRight } from '../types';

function ensureFinite(name: string, value: number): void {
  if (!Number.isFinite(value)) {
    fail('invalid_math_input', `${name} must be finite: ${value}`);
  }
}

function ensurePositive(name: string, value: number): void {
  ensureFinite(name, value);
  if (value <= 0) {
    fail('invalid_math_input', `${name} must be greater than zero: ${value}`);
  }
}

function ensureNonNegative(name: string, value: number): void {
  ensureFinite(name, value);
  if (value < 0) {
    fail('invalid_math_input', `${name} must be non-negative: ${value}`);
  }
}

function validateOptionRight(optionRight: OptionRight): OptionRight {
  if (optionRight !== 'call' && optionRight !== 'put') {
    fail('invalid_math_input', `invalid option right: ${optionRight}`);
  }
  return optionRight;
}

export function price(input: {
  spot: number;
  strike: number;
  years: number;
  rate: number;
  dividendYield: number;
  volatility: number;
  optionRight: OptionRight;
  averageStyle: 'continuous';
}): number {
  ensurePositive('spot', input.spot);
  ensurePositive('strike', input.strike);
  ensurePositive('years', input.years);
  ensureFinite('rate', input.rate);
  ensureFinite('dividendYield', input.dividendYield);
  ensureNonNegative('volatility', input.volatility);
  const optionRight = validateOptionRight(input.optionRight);

  if (input.averageStyle !== 'continuous') {
    fail('unsupported_math_input', `unsupported average_style: ${input.averageStyle}`);
  }

  const sign = optionRight === 'call' ? 1 : -1;
  const variance = input.volatility * input.volatility * input.years / 3;

  if (variance === 0) {
    const meanLevel = input.spot * Math.exp((input.rate - input.dividendYield) * input.years / 2);
    return Math.exp(-input.rate * input.years) * Math.max(sign * (meanLevel - input.strike), 0);
  }

  const meanLn = Math.log(input.spot)
    + (input.rate - input.dividendYield - 0.5 * input.volatility * input.volatility) * input.years / 2;
  const stdDev = Math.sqrt(variance);
  const d1 = (meanLn - Math.log(input.strike) + variance) / stdDev;
  const d2 = d1 - stdDev;
  const expectedAverage = Math.exp(meanLn + 0.5 * variance);

  return Math.exp(-input.rate * input.years) * sign * (
    expectedAverage * normalCdf(sign * d1) - input.strike * normalCdf(sign * d2)
  );
}

File: crates/alpaca-option/src/math/geometric_asian.rs
use crate::error::{OptionError, OptionResult};
use crate::numeric::normal_cdf;
use crate::types::OptionRight;

fn ensure_finite(name: &str, value: f64) -> OptionResult<()> {
    if value.is_finite() {
        Ok(())
    } else {
        Err(OptionError::new(
            "invalid_math_input",
            format!("{name} must be finite: {value}"),
        ))
    }
}

fn ensure_positive(name: &str, value: f64) -> OptionResult<()> {
    ensure_finite(name, value)?;
    if value > 0.0 {
        Ok(())
    } else {
        Err(OptionError::new(
            "invalid_math_input",
            format!("{name} must be greater than zero: {value}"),
        ))
    }
}

fn ensure_non_negative(name: &str, value: f64) -> OptionResult<()> {
    ensure_finite(name, value)?;
    if value >= 0.0 {
        Ok(())
    } else {
        Err(OptionError::new(
            "invalid_math_input",
            format!("{name} must be non-negative: {value}"),
        ))
    }
}

fn parse_option_right(option_right: &str) -> OptionResult<OptionRight> {
    OptionRight::from_str(option_right).map_err(|_| {
        OptionError::new(
            "invalid_math_input",
            format!("invalid option right: {option_right}"),
        )
    })
}

pub fn price(
    spot: f64,
    strike: f64,
    years: f64,
    rate: f64,
    dividend_yield: f64,
    volatility: f64,
    option_right: &str,
    average_style: &str,
) -> OptionResult<f64> {
    ensure_positive("spot", spot)?;
    ensure_positive("strike", strike)?;
    ensure_positive("years", years)?;
    ensure_finite("rate", rate)?;
    ensure_finite("dividend_yield", dividend_yield)?;
    ensure_non_negative("volatility", volatility)?;
    let option_right = parse_option_right(option_right)?;

    if average_style != "continuous" {
        return Err(OptionError::new(
            "unsupported_math_input",
            format!("unsupported average_style: {average_style}"),
        ));
    }

    let sign = match option_right {
        OptionRight::Call => 1.0,
        OptionRight::Put => -1.0,
    };
    let variance = volatility * volatility * years / 3.0;

    if variance == 0.0 {
        let mean_level = spot * ((rate - dividend_yield) * years / 2.0).exp();
        return Ok((-rate * years).exp() * (sign * (mean_level - strike)).max(0.0));
    }

    let mean_ln = spot.ln() + (rate - dividend_yield - 0.5 * volatility * volatility) * years / 2.0;
    let std_dev = variance.sqrt();
    let d1 = (mean_ln - strike.ln() + variance) / std_dev;
    let d2 = d1 - std_dev;
    let expected_average = (mean_ln + 0.5 * variance).exp();

    Ok((-rate * years).exp()
        * sign
        * (expected_average * normal_cdf(sign * d1) - strike * normal_cdf(sign * d2)))
}
